pkg/controller/syncmachineset/gcpactuator.go

```diff
@@ -0,0 +1,126 @@
+package syncmachineset
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+
+	machineapi "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+
+	installgcp "github.com/openshift/installer/pkg/asset/machines/gcp"
+	installtypes "github.com/openshift/installer/pkg/types"
+	installtypesgcp "github.com/openshift/installer/pkg/types/gcp"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+	"github.com/openshift/hive/pkg/gcpclient"
+)
+
+// GCPActuator encapsulates the pieces necessary to be able to generate
+// a list of MachineSets to sync to the remote cluster.
+type GCPActuator struct {
+	client gcpclient.Client
+	logger log.FieldLogger
+	region string
+}
+
+var _ Actuator = &GCPActuator{}
+
+// NewGCPActuator is the constructor for building a GCPActuator
+func NewGCPActuator(gcpCreds []byte, gcpProjectID, region string, logger log.FieldLogger) (*GCPActuator, error) {
+	gcpClient, err := gcpclient.NewClient(gcpProjectID, gcpCreds)
+	if err != nil {
+		logger.WithError(err).Warn("failed to create GCP client with creds in clusterDeployment's secret")
+		return nil, err
+	}
+	actuator := &GCPActuator{
+		client: gcpClient,
+		logger: logger,
+		region: region,
+	}
+	return actuator, nil
+}
+
+// GenerateMachineSets satisfies the Actuator interface and will take a clusterDeployment and return a list of MachineSets
+// to sync to the remote cluster.
+func (a *GCPActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, ic *installtypes.InstallConfig, logger log.FieldLogger) ([]*machineapi.MachineSet, error) {
+	machineSets := []*machineapi.MachineSet{}
+
+	// get image ID for the generated machine sets
+	imageID, err := a.getImageID(cd, logger)
+	if err != nil {
+		a.logger.WithError(err).Warn("failed to find image ID for the machine sets")
+		return nil, err
+	}
+
+	// create the machinesets
+	for _, computePool := range ic.Compute {
+		if computePool.Platform.GCP == nil {
+			computePool.Platform.GCP = &installtypesgcp.MachinePool{}
+		}
+		if len(computePool.Platform.GCP.Zones) == 0 {
+			zones, err := getGCPZones(a.client, a.region)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(zones) == 0 {
+				return nil, fmt.Errorf("fetched 0 zones from cloud provider")
+			}
+			computePool.Platform.GCP.Zones = zones
+		}
+
+		installerMachineSets, err := installgcp.MachineSets(cd.Status.InfraID, ic, &computePool, imageID, computePool.Name, "worker-user-data")
+		if err != nil {
+			logger.WithError(err).Error("failed to generate machinesets")
+			return nil, err
+		}
+
+		for _, ms := range installerMachineSets {
+			machineSets = append(machineSets, ms)
+		}
+	}
+
+	return machineSets, nil
+}
+
+func getGCPZones(client gcpclient.Client, region string) ([]string, error) {
+	zones := []string{}
+
+	// Filter to regions matching '.*<region>.*'
+	zoneFilter := fmt.Sprintf("region eq '.*%s.*'", region)
+	zoneList, err := client.ListComputeZones(gcpclient.ListComputeZonesOptions{Filter: zoneFilter})
```

Comment:
Although I think it is highly unlikely, we still need to handle the possibility that the results will be paged.